draft-ietf-quic-recovery.md

```diff
@@ -162,43 +162,209 @@ latency before a userspace QUIC receiver processes a received packet.
+
+* the latest RTT sample is lower than the SRTT, perhaps due to reordering where
+packet whose ack triggered the Early Retransit process encountered a shorter
+path;
+
+* the SRTT is lower than the latest RTT sample, perhaps due to a sustained
+increase in the actual RTT, but the smoothed SRTT has not yet caught up.
+
+The 1.25 multiplier increases reordering resilience. Implementers MAY experiment
+with using other multipliers, bearing in mind that a lower multiplier reduces
+reordering resilience and increases spurious retransmissions, and a higher
+multipler increases loss recovery delay.
+
+This mechanism is based on Early Retransmit for TCP {{!RFC5827}}. However,
+{{!RFC5827}} does not include the alarm described above. Early Retransmit is
+prone to spurious retransmissions due to its reduced reordering resilence
+without the alarm. This observation led Linux TCP implementers to implement an
+alarm for TCP as well, and this document incorporates this advancement.
+
+
+## Timer-based Detection
+
+Timer-based loss detection implements the spirit of TCP's Tail Loss Probe
+and Retransmission Timeout mechanisms.
+
+### Tail Loss Probe
+
+The algorithm described in this section is an adaptation of the Tail Loss Probe
+algorithm proposed for TCP {{!I-D.draft-dukkipati-tcpm-tcp-loss-probe}}.
+
+A packet sent at the tail is particularly vulnerable to slow loss detection,
+since subsequent packets (and acks) are needed to trigger ack-based
+detection. To ameliorate this weakness of tail packets, the sender schedules an
+alarm when a tail packet is transmitted. When this alarm fires, a Tail Loss
+Probe (TLP) packet is sent to evoke an acknowledgement from the receiver.
+
+The alarm duration, or Probe Timeout (PTO), is set based on the following
+conditions:
+
+* If there is exactly one unacknowledged packet, PTO SHOULD be scheduled for
+  max(2*SRTT, 1.5*SRTT+kDelayedAckTimeout)
+
+* If there are more than one unacknowledged packets, PTO SHOULD be scheduled for
+  max(2*SRTT, 10ms).
+
+* If RTO is earlier, schedule a TLP alarm in its place. That is, PTO SHOULD be
+  scheduled for min(RTO, PTO).
+
+kDelayedAckTimeout is the worst case delayed ACK timer.  When the is exactly one
+unacknowledged packet, the alarm duration includes time for an acknowledgment to
+be received, and additionally, a kDelayedAckTimeout period to compensate for the
+delayed acknowledgment timer at the receiver.
+
+The RECOMMENDED value for kDelayedAckTimeout is 25ms.
+
+A PTO value of at least 2*SRTT ensures that the ACK is overdue. Using a PTO of
+exactly 1*SRTT may generate spurious probes, and 2*SRTT is simply the next
+integral value of RTT.
+
+(TODO: These values of 2 and 1.5 are a bit arbitrary. Reconsider these.)
+
+If the Retransmission Timeout (RTO, {{rto}}) is smaller than the computed PTO,
+then a PTO is scheduled for the smaller RTO period.
+
+To reduce latency, it is RECOMMENDED that the sender send two sequential TLP
```

Comment:
One could potentially read two sequential TLP packets as two packets in a row immediately.  How about wording like "it is RECOMMENDED that the sender allow the TLP alarm to expire twice before setting an RTO alarm."